lib/apiservers/engine/backends/volume.go

```diff
@@ -67,50 +69,119 @@ func NewVolumeModel(volume *models.VolumeResponse, labels map[string]string) *ty
 type Volume struct {
 }
 
+// acceptedVolumeFilters are filters that are supported by VIC
+var acceptedVolumeFilters = map[string]bool{
+	"dangling": true,
+	"name":     true,
+	"driver":   true,
+}
+
+var errPortlayerClient = fmt.Errorf("failed to get a portlayer client")
+
 // Volumes docker personality implementation for VIC
 func (v *Volume) Volumes(filter string) ([]*types.Volume, []string, error) {
 	defer trace.End(trace.Begin("Volume.Volumes"))
 	var volumes []*types.Volume
 
 	client := PortLayerClient()
 	if client == nil {
-		return nil, nil, derr.NewErrorWithStatusCode(fmt.Errorf("Failed to get a portlayer client"), http.StatusInternalServerError)
+		return nil, nil, VolumeInternalServerError(errPortlayerClient)
 	}
 
 	res, err := client.Storage.ListVolumes(storage.NewListVolumesParamsWithContext(ctx).WithFilterString(&filter))
 	if err != nil {
 		switch err := err.(type) {
 		case *storage.ListVolumesInternalServerError:
-			return nil, nil, derr.NewErrorWithStatusCode(fmt.Errorf("error from portlayer server: %s", err.Payload.Message), http.StatusInternalServerError)
+			return nil, nil, VolumeInternalServerError(fmt.Errorf("error from portlayer server: %s", err.Payload.Message))
 		case *storage.ListVolumesDefault:
-			return nil, nil, derr.NewErrorWithStatusCode(fmt.Errorf("error from portlayer server: %s", err.Payload.Message), http.StatusInternalServerError)
+			return nil, nil, VolumeInternalServerError(fmt.Errorf("error from portlayer server: %s", err.Payload.Message))
 		default:
-			return nil, nil, derr.NewErrorWithStatusCode(fmt.Errorf("error from portlayer server: %s", err.Error()), http.StatusInternalServerError)
+			return nil, nil, VolumeInternalServerError(fmt.Errorf("error from portlayer server: %s", err.Error()))
 		}
 	}
 
 	volumeResponses := res.Payload
 
-	log.Infoln("volumes found: ")
+	// Parse and validate filters
+	volumeFilters, err := filters.FromParam(filter)
+	if err != nil {
+		return nil, nil, VolumeInternalServerError(err)
+	}
+	volFilterContext, err := vicfilter.ValidateVolumeFilters(volumeFilters, acceptedVolumeFilters, nil)
+	if err != nil {
+		return nil, nil, VolumeInternalServerError(err)
+	}
+
+	// joinedVolumes stores names of volumes that are joined to a container
+	// and is used while filtering the output by dangling (dangling=true should
+	// return volumes that are not attached to a container)
+	joinedVolumes := make(map[string]struct{})
+
+	if volumeFilters.Include("dangling") {
+		// If the dangling filter is specified, gather required items beforehand
+
+		// Get all containers from the portlayer and use their metadata to obtain
+		// non-dangling (joined) volumes
+		conts, err := allContainers()
+		if err != nil {
+			return nil, nil, VolumeInternalServerError(err)
+		}
+
+		var s struct{}
+		for i := range conts {
+			for _, vol := range conts[i].VolumeConfig {
+				joinedVolumes[vol.Name] = s
+			}
+		}
+	}
+
+	log.Infoln("volumes found:")
```

Comment:
Those have been there since 98378fc23eae0fa9f103b71ea34aa3be944e35b1 and seem like they'd be helpful to tell which volumes got filtered out.